README
NAME
    VS::RuleEngine - Generic rule based processing engine

SYNOPSIS
    A simple example which just randomizes a number 10 times and outputs it
    to screen if it was under 500.

      use VS::RuleEngine::Constants;
      use VS::RuleEngine::Declare;
      
  my $i = 0;
      
  my $engine = engine {
          prehook "check_i" => does {
              $i++;
              
          return KV_ABORT if $i == 10;
              return KV_CONTINUE;
          };
          
      input "random" => does {
              return int(rand(1000));
          };
          
      rule "under_500" => does {
              my $input = $_[KV_INPUT];
              return KV_MATCH if $input->get("random") < 500;
              return KV_NO_MATCH;
          };
          
      action "mark_under_500" => does {
              my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
              $local->set("output" => $input->get("random") . " was under 500");
          };

          run "mark_under_500" => when "under_500";
          
      output "print" => does {
              my $local = $_[KV_LOCAL];
              my $output = $local->get("output");
              print $output, "\n" if defined $output;
          };
      }

    A fictional spam filter

      use VS::RuleEngine::Constants;
      use VS::RuleEngine::Declare;
      
  my $engine = engine {
          prehook "has_more_email" => instanceof "MyApp::HasMoreEmails";
          
      # Returns an Email::Simple instance
          input "email" => instanceof "MyApp::GetEmail";
          
      input "from" => does {
                my $input = $_[KV_INPUT];
                return $input->get("email")->header("From");
          };
          
      rule "is_spam" => instanceof "MyApp::SpamDetector";
          
      rule "is_from_boss" => does {
              my $input = $_[KV_INPUT];
              
          if ($input->get("from") =~ /theboss@company\.com/) {
                  return KV_MATCH;
              }
              
          return KV_NO_MATCH;
          };
          
      action "mark_for_deletion" => does {
              my $local = $_[KV_LOCAL];
              $local->set("delete-email" => 1);
          };
          
      run "mark_for_deletion" => when qw(is_spam is_from_boss);
          
      posthook "process_mail" => does {
              my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
              
          if ($local->get("delete-mail") == 1) {
                  MyApp::EmailHandler->delete($input->get("email"));
              }
              
          return KV_CONTINUE;
          };
      };
      
  $engine->run();

DESCRIPTION
    VS::RuleEngine is a generic rule based processing engine. Processing is
    done while neither any pre nor post iteration hook aborts processing.
    Each engine supports multiple pre- and posthooks, inputs, output,
    actions and rules. An action is attached to one or several rules and is
    executed if any of the rules matches. Input is processed when needed and
    only once per iteration.

    Multiple engines can be run in parallel in a runloop. Runloops can be
    ran until no more data is available for processing or by a single step
    at a time. This way multiple runloops may be executed simultaneously.

USAGE
    VS::RuleEngine has a declarative interface via VS::RuleEngine::Declare
    which is the simplest way to define engines. The synopsis shows an
    exmaple that could be implemented (with reservations for syntax errors).

  Declarative interface
    VS::RuleEngine::Declare

  Loading engines from other sources
    Currently creation of engines are only possible using either the
    declarative interface or by creating "VS::RuleEngine::Engine"-objects
    directly. Future releases will allow loading of engine declarations via
    XML and other formats.

  Runloops
    Runloops transforms engines into something executable and runs them.

    VS::RuleEngine::Runloop

  Inputs, Outputs, Rules, Hooks, Actions
    The components that makes up an engine.

    VS::RuleEngine::Input, VS::RuleEngine::Output, VS::RuleEngine::Rule,
    VS::RuleEngine::Hook, VS::RuleEngine::Action

BUGS AND LIMITATIONS
    Please report any bugs or feature requests to
    "bug-vs-ruleengine@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org>.

AUTHOR
    Claes Jakobsson "<claesjac@cpan.org>"

LICENCE AND COPYRIGHT
    Copyright (c) 2007 - 2008, Versed Solutions "<info@versed.se>". All
    rights reserved.

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 0.08 - 2008-01-01]

* Renamed to VS::RuleEngine
* Added KV_SKIP return code for rule to abort further processing of the loop and goto processing
  of posthooks directly. 
* Fixed bug caused because iterator is not reset if exiting a each prematurely [BUG]
* Fixed a few doc typos
* Added Kvasir::Action::Set(Local|Global) which sets key/values to the local and global 
  runloop objects

[Changes for 0.07 - 2007-10-08]

* with_args argument is now passed as a hash instead of a hash reference to entities
* Added 'defaults' and 'with_defaults' to declarative interface
* Added default argument sets to entities
* Corrected typo (post_hook => posthook) in Kvasir synopsis [DOC]
* Corrected POD names in InputUnderThreshold to actually match that module [DOC]

[Changes for 0.06 - 2007-10-02]

* Don't install Test::Kvasir::*
* Doesn't require the non-existing Perl 5.6.20 (5.00602) anymore. Spotted by David Cantrell [BUG]
* Added generic rules InputMatchesRegexp, InputOverThreshold and InputUnderThreshold

[Changes for 0.05 - 2007-10-01]

* Don't try to require a package if it's already defined [BUG]
* is_existing_package in Kvasir::Util should now work correctly [BUG]
* Corrected calling posthooks so that their argument positions are correct [BUG]
* Specifying a default global is now possible either via add_engine for Runloop or via run for Engine

[Changes for 0.04 - 2007-09-16]

* add_* in Engine now also accepts objects that'll be used to the specific item. Declarative interface
  is supports this using 'input => $object' etc.
* Added POD to Kvasir/Engine/* to make doc tests not complain that did does not contain any POD [DOC]

[Changes for 0.03 - 2007-09-16]

* Added some synopsises and better documentation [DOC]
* Removed some double docs from Engine.pm and cleaned it up a bit [DOC]
* Kvasir/Engine/*.pm doesn't set package anymore to avoid confusion and making them mixins [BUG/DOC]
 
[Changes for 0.02 - 2007-09-16]

* Inputs are now reset before each iteration [BUG]
* Added another example in synopsis so that peopledon't think this is a spam filter. [DOC]
* Updated documentation with some fixes. [DOC]
  
[Changes for 0.01 - 2007-09-15]

* Initial release to CPAN


lib/VS/RuleEngine/Runloop.pm
package VS::RuleEngine::Runloop;

use strict;
use warnings;

use Carp qw(croak);
use List::Util qw(first);
use Scalar::Util qw(refaddr blessed);

use VS::RuleEngine::Constants;

my %Engine;
my %Runloop;
my %Initialized;

sub new {
    my ($pkg) = @_;
    my $self = bless \do { my $v; }, $pkg;
    $$self = refaddr $self;

    $Engine{$$self} = [];
    $Runloop{$$self} = [];
    $Initialized{$$self} = 0;

    return $self;
}

sub DESTROY {
    my $self = shift;
    delete $Engine{$$self};
    delete $Initialized{$$self};
    delete $Runloop{$$self};
}

sub add_engine {
    my ($self, $engine, $global) = @_;
    
    croak "Engine is undefined" if !defined $engine;
    croak "Engine is not a VS::RuleEngine::Engine instance" if !(blessed $engine && $engine->isa("VS::RuleEngine::Engine"));
    
    my $engines = $Engine{$$self};
    if (!first { $_->[0] == $engine } @$engines) {
        push @$engines, [$engine, $global];
    }
    else {
        croak "Engine already exists";
    }    
}

sub init {
    my ($self) = @_;
    
    return if $Initialized{$$self};
    
    for my $engine (@{$Engine{$$self}}) {
        my $runloop = _mk_runloop($engine->[0], $engine->[1]);
        $self->_register_runloop($runloop);
    }
}

sub _mk_runloop {
    my $engine = shift;
    my $global = shift;
    
    my @rules = @{$engine->_rule_order};

    my %action_map  = map {
        my $actions = $engine->_get_rule_actions($_);
        $_ => [@{$actions}];
    } @rules;
    
	my %rules       = map {
	    my $rule = $engine->_get_rule($_);
	    my $rule_obj = $rule->instantiate($engine);
	    $_ => $rule_obj;
    } @rules;
    
    my %actions     = map {
        my $action = $engine->_get_action($_);
        my $action_obj = $action->instantiate($engine);
        $_ => $action_obj;
    } $engine->actions;
    
	my @pre_hooks	= map { $_->instantiate($engine); } map { $engine->_get_hook($_) } @{$engine->_pre_hooks};
	my @post_hooks	= map { $_->instantiate($engine); } map { $engine->_get_hook($_) } @{$engine->_post_hooks};

	my $inputs		= $engine->_input_handler;
	my @outputs		= map { $_->instantiate($engine); } map { $engine->_get_output($_) } sort $engine->outputs;

	$global = $global || VS::RuleEngine::Data->new();
	    
	$inputs->set_global($global);

	my $runloop = sub {
	    $inputs->_clear();
	    
		my $local = VS::RuleEngine::Data->new();
		
		$inputs->set_local($local);
		
		# Process all pre hooks
		for my $hook (@pre_hooks) {
			my $result = $hook->invoke($inputs, $global, $local);
			return KV_ABORT if $result == KV_ABORT;
		}
		
		# Run rules until we find a matching rule
		my $match;
		my $skip = 0;
		
		PROCESS_RULES: for (@rules) {
		    my $rule = $rules{$_};
			my $result = $rule->evaluate($inputs, $global, $local);
			$skip = 1, last PROCESS_RULES if $result == KV_SKIP;
			$match = $_, last PROCESS_RULES if $result == KV_MATCH;
		}
        
        # Run all actions
        if (!$skip && $match) {
            $local->set('VS::RuleEngine/matchingRule' => $match);
            my $actions = $action_map{$match};
            for (@$actions) {
                my $action = $actions{$_};
                $action->perform($inputs, $global, $local);
            }
        }
        
        # All outputs are always called
        if (!$skip) {
		    PROCESS_OUTPUT: for my $output (@outputs) {
			    $output->process($inputs, $global, $local);
		    }
		}
		
		# Process all post hooks
		for my $hook (@post_hooks) {
			my $result = $hook->invoke($inputs, $global, $local);
			return KV_ABORT if $result == KV_ABORT;
		}
		
		return KV_CONTINUE;
	};
	
	return $runloop;
}

sub _register_runloop {
    my ($self, $runloop) = @_;
    
    my $runloops = $Runloop{$$self};
    if (!first { $_ == $runloop } @$runloops) {
        push @$runloops, $runloop;
    }
    else {
        croak "Runloop already registered";
    }
}

sub _unregister_runloop {
    my ($self, $runloop) = @_;
    
    my $runloops = $Runloop{$$self};
    my @runloops = grep { $_ != $runloop } @$runloops;
    $Runloop{$$self} = \@runloops;
    
}

sub step {
    my $self = shift;
    
    my $runloops = $Runloop{$$self};
    
    if (@$runloops) {
        for my $runloop (@$runloops) {
            my $status = $runloop->();
            if ($status == KV_ABORT) {
                $self->_unregister_runloop($runloop);
            }
        }
    }
    
    return scalar @$runloops;
}

sub run {
    my $self = shift;
    
    $self->init();
    
    RUNLOOP: while(1) {
        last RUNLOOP if $self->step() == 0;
    }    
}

1;
__END__

=head1 NAME

VS::RuleEngine::Runloop - Runs engine(s)

=head1 SYNOPSIS

  use VS::RuleEngine::Runloop;
  
  my $engine1 = MyApp::Engine->get_engine();
  my $engine2 = MyApp::Engine->get_another_engine();
  
  my $runloop = VS::RuleEngine::Runloop->new();
  $runloop->add_engine($engine1);
  $runloop->add_engine($engine2);

  # Run the two engines until there's no more processing
  # to be done in either of them
  $runloop->run();
  
=head1 DESCRIPTION

This class converts engine descriptions (C<VS::RuleEngine::Engine>-instances) into something 
runnable and executes them.

If multiple engines are defined in a single runloop execution will continue until no
engines have anything ore to process. Engines that report they are done processing are 
removed from the runloop.

=head1 INTERFACE

=head2 CLASS METHODS

=over 4

=item new

Creates a new runloop instance.

=back

=head2 INSTANCE METHODS

=over 4

=item add_engine ( ENGINE [, GLOBALS] )

Adds an engine to the runloop. An optional VS::RuleEngine::Data instance can be passed 
as global data for the engine. If omitted a empty VS::RuleEngine::Data instance will be 
created and used as global.

=item init

Initializes the runloop. Must be called before C<step> in order for 
the runloop to work. It called automaticly by C<run>.

=item step

Performs one iteration in the runloop. Returns the number of engines 
that are still in the runloop. When there is no more work to be done it returns 0.

=item run

Runs the engine until there is no more work to perform.

=back

=cut




lib/VS/RuleEngine.pm
package VS::RuleEngine;

use strict;
use warnings;

our $VERSION = "0.08";

1;
__END__

=head1 NAME

VS::RuleEngine - Generic rule based processing engine

=head1 SYNOPSIS

A simple example which just randomizes a number 10 times and outputs it to screen if it was under 500.

  use VS::RuleEngine::Constants;
  use VS::RuleEngine::Declare;
  
  my $i = 0;
  
  my $engine = engine {
      prehook "check_i" => does {
          $i++;
          
          return KV_ABORT if $i == 10;
          return KV_CONTINUE;
      };
      
      input "random" => does {
          return int(rand(1000));
      };
      
      rule "under_500" => does {
          my $input = $_[KV_INPUT];
          return KV_MATCH if $input->get("random") < 500;
          return KV_NO_MATCH;
      };
      
      action "mark_under_500" => does {
          my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
          $local->set("output" => $input->get("random") . " was under 500");
      };

      run "mark_under_500" => when "under_500";
      
      output "print" => does {
          my $local = $_[KV_LOCAL];
          my $output = $local->get("output");
          print $output, "\n" if defined $output;
      };
  }

A fictional spam filter

  use VS::RuleEngine::Constants;
  use VS::RuleEngine::Declare;
  
  my $engine = engine {
      prehook "has_more_email" => instanceof "MyApp::HasMoreEmails";
      
      # Returns an Email::Simple instance
      input "email" => instanceof "MyApp::GetEmail";
      
      input "from" => does {
            my $input = $_[KV_INPUT];
            return $input->get("email")->header("From");
      };
      
      rule "is_spam" => instanceof "MyApp::SpamDetector";
      
      rule "is_from_boss" => does {
          my $input = $_[KV_INPUT];
          
          if ($input->get("from") =~ /theboss@company\.com/) {
              return KV_MATCH;
          }
          
          return KV_NO_MATCH;
      };
      
      action "mark_for_deletion" => does {
          my $local = $_[KV_LOCAL];
          $local->set("delete-email" => 1);
      };
      
      run "mark_for_deletion" => when qw(is_spam is_from_boss);
      
      posthook "process_mail" => does {
          my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
          
          if ($local->get("delete-mail") == 1) {
              MyApp::EmailHandler->delete($input->get("email"));
          }
          
          return KV_CONTINUE;
      };
  };
  
  $engine->run();

=head1 DESCRIPTION

VS::RuleEngine is a generic rule based processing engine. Processing is done while neither any pre nor post iteration hook 
aborts processing. Each engine supports multiple pre- and posthooks, inputs, output, actions and rules. An action is 
attached to one or several rules and is executed if any of the rules matches. Input is processed when needed and only 
once per iteration.

Multiple engines can be run in parallel in a runloop. Runloops can be ran until no more data is available for processing 
or by a single step at a time. This way multiple runloops may be executed simultaneously.

=head1 USAGE

VS::RuleEngine has a declarative interface via L<VS::RuleEngine::Declare> which is the simplest way to define engines. The synopsis shows 
an exmaple that could be implemented (with reservations for syntax errors).

=head2 Declarative interface

L<VS::RuleEngine::Declare>

=head2 Loading engines from other sources

Currently creation of engines are only possible using either the declarative interface or by 
creating C<VS::RuleEngine::Engine>-objects directly. Future releases will allow loading of engine declarations 
via XML and other formats.

=head2 Runloops

Runloops transforms engines into something executable and runs them.

L<VS::RuleEngine::Runloop>

=head2 Inputs, Outputs, Rules, Hooks, Actions

The components that makes up an engine.

L<VS::RuleEngine::Input>, L<VS::RuleEngine::Output>, L<VS::RuleEngine::Rule>, L<VS::RuleEngine::Hook>, L<VS::RuleEngine::Action>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-vs-ruleengine@rt.cpan.org>, 
or through the web interface at L<http://rt.cpan.org>.

=head1 AUTHOR

Claes Jakobsson C<< <claesjac@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007 - 2008, Versed Solutions C<< <info@versed.se> >>. All rights reserved.

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
